### PR TITLE
Trial method of speeding up jest tests

### PR DESCRIPTION
**Pull request by zephraph**
## Goal

Increase the overall iteration speed of PRs without decreasing the availability of CI. 

## Approach

- For PRs without dependency changes utilize jest's [changedSince]() mechanism to only run the subset of tests that might've changed for that PR. 
- For PRs with dependency changes, run the full suite of tests
- For all master builds, run the full suite of tests
- This functionality is **disabled by default** and must be opted into by setting the `only_test_changed` 

## Tradeoffs

- It's explicit that for PRs we will _not_ be running all tests (unless a dependency updates). Only tests related to what jest sees has having changed will be tested. This trades faster iteration cycles for more potential upstream failures.

## Usage

To enable this feature a consuming project use the following config:

```yaml
- yarn/jest:
    args: --runInBand
    only_test_changed: true
    run_all_tests_if_these_files_change: src/setup_jest.ts .babelrc patches/*
```

`only_test_changed` is the flag to opt in to the feature. `run_all_tests_if_these_files_change` is a list of files to check for changes before triggering a full test suite. 

_This will only run partial tests on non-master branches_.

canary: artsy/yarn@dev:55f6486eeacc499426acd697242595df

### CODE REVIEW

**Comment by eessex** [comment]
It looks like this method is trying to speed up tests by not running them all, which worries me. I wonder if we could experiment with some other strategies first, using Circle to 'glob' tests into parallel containers was one that I've been hoping to try for Reaction: https://circleci.com/docs/2.0/parallelism-faster-jobs/#using-the-circleci-cli-to-split-tests

**Comment by zephraph** [comment]
I hear your concerns and share them. To somewhat mitigate those concerns it only runs a reduced set of changes on PRs and only when there aren't any global changes (like dependency updates) that it can't inspect. The actual logic of running only what has changed is actually built into jest. 

Master will still be gated by a full test run. I think this tradeoff trades faster iteration times with potential delays of releases if the process breaks down. That said, I believe it'd be worth trying and if we have issues we can either tweak it or disable it. I'd even be willing to update this to be toggled by an env var. 

I think your idea of parallelizing tests is a good one. Even with this change, I'd definitely like to see that on master. My concern there, is that if every PR is using parallel tests then we could significantly use up our available CI containers and end up w/ higher wait times. 

The last part of the puzzle is actually optimizing tests. We have some that are 10+ seconds (for a single test). That's less than ideal.

**Comment by eessex** [comment]
Can we limit this to renovate and dependabot PRs or some kind of flag for now? I'm nervous not knowing the full range of apps that use this orb to run their tests.

Re optimizing tests, we generate metadata on tests but don't yet store it-- we should enable `store_test_results` to get data on our slowest tests to target: https://circleci.com/docs/2.0/collect-test-data/. I will create a separate ticket for this.

If the tests are actually fast, I'm interested to try using 2 containers for Reaction.

**Comment by zephraph** [comment]
To be clear, this PR isn't done. I've just been experimenting. 

The final version, if we decide to move forward with it, would be opt in only.

**Comment by zephraph** [comment]
Oh, and @eessex we do store test results: https://app.circleci.com/jobs/github/artsy/reaction/52617/parallel-runs/0/steps/0-106

File: src/yarn/yarn.yml [hunk 1]
```diff
@@ -117,11 +119,66 @@ jobs:
...
+                env-vars: SLACK_WEBHOOK
+      - setup
+      - when:
+          condition: << parameters.only_test_changed >>
+          steps:
+            - run:
+                name: jest
+                command: |
+                  important_files_modified() {
+                    FILES="yarn.lock << parameters.important_files >>"
```
Comment (damassi) [inline src/yarn/yarn.yml]: 👍

**Comment by ds300** [comment]
@zephraph was playing around with `--changedSince` and noticed that changes to files in `__mocks__` folders also don't trigger re-testing.

**Comment by zephraph** [comment]
Interesting... I suppose that'd make sense from one perspective, those mocks aren't necessarily directly in the module mappings. Could just add those to the file watch list.

**Comment by zephraph** [comment]
The way this PR is currently configured, it's entirely a no-op and only makes changes in an opt-in basis. I'm going to ship it out for a trial run.

**Pull request merged**